### PR TITLE
Check for namespace-qualified packages in repo_for_pkg

### LINKED ISSUE #5754: spack does not handle multiple repos any more
I had recently updated develop. I was using an additional repo.
 Now this seem to prevent spack to work:

To reproduce:
cd /tmp
git clone https://github.com/LLNL/spack.git 
cd spack
source share/spack/setup-env.sh 
spack spec zlib
---------- WORK as expected ----------
spack repo create /tmp/goofy
spack repo add /tmp/goofy
spack spec zlib
==> Error: Invalid namespace for goofy repo: builtin

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes #5754

Previously when RepoPath.repo_for_pkg was invoked with a string,
it did not check if the string included a namespace. Any
namespace-qualified package provided as a string would not be found
(at which point the behavior was to return the highest-precedence
repository).

### CODE REVIEW

**Comment by scheibelp** [comment]
Recently, https://github.com/LLNL/spack/pull/5716 added calls to `Spec.package_class` which invokes `RepoPath.repo_for_pkg` with a namespace-qualified name. Other instances of `Spec.package_class` would have broken before but adding the calls in #5716 made it "more likely" to encounter an issue when there were multiple repositories.

**Comment by scheibelp** [comment]
I'll be able to add a unit test for this later today.

File: lib/spack/spack/repository.py [hunk 1]
```diff
@@ -541,19 +541,24 @@ def repo_for_pkg(self, spec):
...
-                    raise UnknownNamespaceError(spec.namespace)
-                return self.by_namespace[fullspace]
+            namespace = spec.namespace
             name = spec.name
-
         else:
             # handle strings directly for speed instead of @_autospec'ing
-            name = spec
+            try:
+                namespace, name = spec.split('.')
```
Comment (tgamblin) [inline lib/spack/spack/repository.py]: Namespaces can be nested, so I think you want:

```
namespace, _, name = spec.rpartition('.')
```

to do this right.  This is what `Repo` uses.  `namespace` will be empty if there is no namespace.

**Comment by scheibelp** [comment]
@tgamblin added tests and made suggested update.

To make this more-easily tested (in particular separable from global state), https://github.com/LLNL/spack/pull/5787/commits/ef360f540f89d798b0185f34f00d0d83ec954215 replaced the `spec.virtual` call in `Repo.get`

**Comment by scheibelp** [comment]
@yee379 as far as I can tell I can't reply directly to your comment at https://github.com/LLNL/spack/commit/65b38764aedf56192568f5d49eb9a9c87257761b#commitcomment-25042386 but this PR should fix your issue (which is likely another case of #5754).

File: lib/spack/spack/test/repo.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import pytest
+
+
+# Unlike the repo_path fixture defined in conftest, this has a test-level
+# scope rather than a session level scope, since we want to edit the
+# given RepoPath
+@pytest.fixture()
+def test_repo():
+    return spack.repository.RepoPath(spack.mock_packages_path)
```
Comment (alalazo) [inline lib/spack/spack/test/repo.py]: Hmmm, `pytest` looks for methods and functions [that start with `test_`](https://docs.pytest.org/en/latest/goodpractices.html#test-discovery) and consider them as tests. Not sure if creating a fixture that matches this pattern is asking for troubles...

**Comment by scheibelp** [comment]
> Hmmm, pytest looks for methods and functions that start with test_ and consider them as tests. Not sure if creating a fixture that matches this pattern is asking for troubles...

I have edited the name - thanks for pointing that out!

**Pull request merged**